examples/replit/main.cpp

```diff
@@ -0,0 +1,850 @@
+        }
+      }
+    }
+  }
+
+  if (best_segmentations_scores.back() ==
+      -std::numeric_limits<float>::infinity()) {
+    return std::make_tuple(std::vector<std::size_t>{0}, 0.0);
+  }
+
+  float score = best_segmentations_scores.back();
+  int start = best_segmentations_starts.back();
+  int end = word.length();
+  std::vector<std::size_t> tokens;
+  while (start != 0) {
+    const auto token_id = model.at(word.substr(start, end - start)).first;
+    tokens.insert(tokens.begin(), token_id);
+    int next_start = best_segmentations_starts[start];
+    end = start;
+    start = next_start;
+  }
+  const auto token_id = model.at(word.substr(start, end - start)).first;
+  tokens.insert(tokens.begin(), token_id);
+  return std::make_tuple(tokens, score);
+}
+
+auto replit_tokenizer_load(replit_tokenizer &tokenizer, std::istream &fin,
+                           int max_vocab_size) -> bool {
+
+  for (std::size_t i = 0; i < max_vocab_size; i++) {
+
+    uint32_t len;
+    fin.read((char *)&len, sizeof(len));
+
+    std::string word;
+    word.resize(len);
+    fin.read((char *)word.data(), len);
+
+    float score;
+    fin.read((char *)&score, sizeof(score));
+
+    tokenizer.piece_map[word] = std::make_pair(i, -score);
+    tokenizer.raw_vocab.id_to_token[i] = word;
+  }
+
+  return true;
+}
+
+std::string replace_all(const std::string &str,    // where to work
+                        const std::string &find,   // substitute 'find'
+                        const std::string &replace //      by 'replace'
+) {
+  using namespace std;
+  string result;
+  size_t find_len = find.size();
+  size_t pos, from = 0;
+  while (string::npos != (pos = str.find(find, from))) {
+    result.append(str, from, pos - from);
+    result.append(replace);
+    from = pos + find_len;
+  }
+  result.append(str, from, string::npos);
+  return result;
+}
+
+std::string ws_symbol = "\342\226\201";
+auto replit_tokenizer_tokenize(replit_tokenizer &tokenizer,
+                               const std::string &text)
+    -> std::vector<std::size_t> {
+  std::vector<std::size_t> tokens;
+  auto normalized_text = replace_all(text, " ", ws_symbol);
+  normalized_text = replace_all(normalized_text, "\\n", "\n");
+  auto tokenized = encode_word(normalized_text, tokenizer.piece_map);
+
+  return tokenized.first;
+}
+
+auto replit_tokenizer_detokenize(replit_tokenizer &tokenizer,
+                                 const std::vector<std::size_t> &tokens)
+    -> std::string {
+  std::string text;
+  for (auto token : tokens) {
+    text += tokenizer.raw_vocab.id_to_token[token];
+  }
+  auto denormalized_text = replace_all(text, ws_symbol, " ");
+  return denormalized_text;
+}
+
+// no defaults for now
+struct mpt_hparams {
+  int32_t d_model = 0;
+  int32_t max_seq_len = 0;
+  int32_t n_heads = 0;
+  int32_t n_layers = 0;
+  int32_t n_vocab = 0;
+  int32_t ftype = 0;
+};
+
+struct replit_layer {
+  // pre normalization
+  struct ggml_tensor *ln_1_weight;
```

Comment:
The style for pointers and reference is "space before and after":

```suggestion
  struct ggml_tensor * ln_1_weight;
```